### PR TITLE
Fsoc 24

### PR DESCRIPTION
**Pull request by GDW1**
## Description

Every 24 hours, FSOC will check if there is a update available. It will create a temp file that holds the most recent timestep

## Type of Change

- [ ] Bug Fix
- [x] New Feature
- [ ] Breaking Change
- [ ] Refactor
- [ ] Documentation
- [ ] Other (please describe)

## Checklist

- [x] I have read the [contributing guidelines](/CONTRIBUTING.md)
- [x] Existing issues have been referenced (where applicable)
- [x] I have verified this change is not present in other open pull requests
- [ ] Functionality is documented
- [x] All code style checks pass
- [ ] New code contribution is covered by automated tests
- [ ] All new and existing tests pass

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -222,6 +229,116 @@ func preExecHook(cmd *cobra.Command, args []string) {
...
+	newestVersionChan := make(chan string)
+	go func() {
+		err := getVersion(newestVersionChan)
+		if err != nil {
+			log.Fatalf(err.Error())
+		}
+	}()
+	var newestVersion = <-newestVersionChan
+	if compareVersion(newestVersion) {
+		log.Warnf("There is a newer version of FSOC available, please upgrade to version %s", newestVersion)
```
Comment (pnickolov) [inline cmd/root.go:289]: it may be nice to print both the current and the latest version. See pip, for example here https://stackoverflow.com/questions/46288847/how-to-suppress-pip-upgrade-warning
(although a single-line message is better; consider also using WithFields, so that the message can be parsed programmatically and the version extracted)

File: cmd/root.go [hunk 2]
```diff
@@ -222,6 +229,116 @@ func preExecHook(cmd *cobra.Command, args []string) {
...
+	}
+}
+
+func getVersion(ver chan string) error {
```
Comment (pnickolov) [inline cmd/root.go:293]: This should probably be getLatestVersion(). It will also make sense to separate the channel/async functionality from the specific functionality (unless you're relying on the client Get async completion, which is not what I see here.)

File: cmd/root.go [hunk 3]
```diff
@@ -222,6 +229,116 @@ func preExecHook(cmd *cobra.Command, args []string) {
...
+			return http.ErrUseLastResponse
+		},
+	}
+	resp, err := client.Get("https://github.com/cisco-open/fsoc/releases/latest")
+	if err != nil {
+		return err
+	}
+
+	split := strings.Split(resp.Header.Get("Location"), "/")
+	ver <- split[len(split)-1]
```
Comment (pnickolov) [inline cmd/root.go:305]: This is using data received from the outside and must be sanitized. Please check the obvious (len >= 1, before indexing). Also, consider checking at least the valid charset (per https://semver.org/#backusnaur-form-grammar-for-valid-semver-versions), very loosely (e.g., [A-Za-z0-9.-]), esp. no control characters

File: cmd/root.go [hunk 4]
```diff
@@ -222,6 +229,116 @@ func preExecHook(cmd *cobra.Command, args []string) {
...
+	return nil
+}
+
+func compareVersion(newestVersion string) bool { // Returns if the parameter is a newer version than the installed one
```
Comment (pnickolov) [inline cmd/root.go:309]: For parsing and comparisons, please use a proper semver parser/comparer. We already use the github.com/Masterminds/semver package, so this should be good enough. Ensure that failure to parse is also considered/handled in an intentional manner.

File: cmd/root.go [hunk 5]
```diff
@@ -222,6 +229,116 @@ func preExecHook(cmd *cobra.Command, args []string) {
...
+func parseVersion(version string) (uint64, uint64, uint64) {
+	version = version[1:]
+	versionSections := strings.Split(version, ".")
+	major, _ := strconv.ParseUint(versionSections[0], 10, 32)
+	minor, _ := strconv.ParseUint(versionSections[1], 10, 32)
+	patch, _ := strconv.ParseUint(versionSections[2], 10, 32)
+	return major, minor, patch
+}
+
+func postExecHook(cmd *cobra.Command, args []string) {
```
Comment (pnickolov) [inline cmd/root.go:336]: Why is this here/on post. Ideally, we should set the timestamp as soon as we complete the check successfully (regardless  of whether there is a newer version or not; maybe even regardless of whether we got the latest version and it parsed OK)

**Review by pnickolov — CHANGES_REQUESTED** [review]

File: cmd/version/version.go [hunk 6]
```diff
@@ -64,3 +68,62 @@ func displayVersion(cmd *cobra.Command) {
 		Detail:  true,
 	})
 }
+
+func GetLatestVersion() (string, error) {
+	client := &http.Client{
```
Comment (pnickolov) [inline cmd/version/version.go:73]: consider adding a clarifying comment, something like 
```
// create a http client that will not follow redirect
```

File: cmd/version/version.go [hunk 7]
```diff
@@ -64,3 +68,62 @@ func displayVersion(cmd *cobra.Command) {
...
+		CheckRedirect: func(req *http.Request, via []*http.Request) error { // no redirect
+			return http.ErrUseLastResponse
+		},
+	}
+	resp, err := client.Get("https://github.com/cisco-open/fsoc/releases/latest")
```
Comment (pnickolov) [inline cmd/version/version.go:78]: please move the string constant to a const

File: cmd/version/version.go [hunk 8]
```diff
@@ -64,3 +68,62 @@ func displayVersion(cmd *cobra.Command) {
...
+	if err != nil {
+		return "", err
+	}
+	split := strings.Split(resp.Header.Get("Location"), "/")
```
Comment (pnickolov) [inline cmd/version/version.go:82]: you may want to give a comment with an example of what the redirect string is expected to look like; will make understanding this easier

File: cmd/version/version.go [hunk 9]
```diff
@@ -64,3 +68,62 @@ func displayVersion(cmd *cobra.Command) {
...
+func CheckForUpdate(versionChannel chan *semver.Version) {
+	defer func() {
+		r := recover()
+		if r != nil {
+			log.Warnf("Error occurred while checking version, continuing")
+		}
+	}()
+	log.Infof("Checking for newer version of FSOC")
+	newestVersion, err := GetLatestVersion()
+	log.Infof("Latest version available: %s", newestVersion)
```
Comment (pnickolov) [inline cmd/version/version.go:98]:
```suggestion
	log.Infof("Latest fsoc version available: %s", newestVersion)
```
Remember that the user is not thinking about fsoc itself but about the command they ran. Let's make sure any logged messages are clearly stating what operation they pertain to.

Comment (pnickolov) [reply cmd/version/version.go:98]: Also, please consider using WithField (there is a version for a single field, easier to use than WithFields), so that the field is programmatically parseable in the log

File: cmd/version/version.go [hunk 10]
```diff
@@ -64,3 +68,62 @@ func displayVersion(cmd *cobra.Command) {
...
+	if err != nil {
+		log.Warnf(err.Error())
+	}
+	newestVersionSemVar, err := semver.NewVersion(newestVersion) // This panics, so we need to be ready to recover
```
Comment (pnickolov) [inline cmd/version/version.go:102]: It shouldn't panic, according to docs: https://pkg.go.dev/github.com/Masterminds/semver/v3#NewVersion
(but if it does, then OK, may be good to open an issue on it)

File: cmd/version/version.go [hunk 11]
```diff
@@ -64,3 +68,62 @@ func displayVersion(cmd *cobra.Command) {
...
+	if err != nil {
+		log.Warnf(err.Error())
```
Comment (pnickolov) [inline cmd/version/version.go:104]: You may want to provide some more info, incl. a printout of the version that could not be parsed

File: cmd/version/version.go [hunk 12]
```diff
@@ -64,3 +68,62 @@ func displayVersion(cmd *cobra.Command) {
...
+	}
+	versionChannel <- newestVersionSemVar
+}
+
+func CompareAndLogVersions(newestVersionSemVar *semver.Version) {
+	currentVersion := GetVersion()
+	currentVersionSemVer := ConvertVerToSemVar(currentVersion)
+	newerVersionAvailable := currentVersionSemVer.Compare(newestVersionSemVar) == -1
+	var debugFields = log.Fields{"newerVersionAvailable": newerVersionAvailable, "oldVersion": currentVersionSemVer.String(), "newVersion": newestVersionSemVar.String()}
```
Comment (pnickolov) [inline cmd/version/version.go:113]: I don't think we should include the newerVersionAvailable field. This should be--and is--part of the message: a newer version is available.
Also, instead of "old" and "new" version field names, it will be more helpful to call them "current" (i.e., this is what your current fsoc that is running is) and "latest" (i.e., this is what is available / you should run)

File: cmd/version/version.go [hunk 13]
```diff
@@ -64,3 +68,62 @@ func displayVersion(cmd *cobra.Command) {
...
+
+	if newerVersionAvailable {
+		log.WithFields(debugFields).Warnf("There is a newer version of FSOC available, please upgrade from version %s to version %s", currentVersionSemVer.String(), newestVersionSemVar.String())
+	} else {
+		log.WithFields(debugFields)
```
Comment (pnickolov) [inline cmd/version/version.go:118]: I don't think we need to log this; but it we do, I propose we just log the version values

File: cmd/version/version.go [hunk 14]
```diff
@@ -64,3 +68,62 @@ func displayVersion(cmd *cobra.Command) {
...
+}
+
+func CompareAndLogVersions(newestVersionSemVar *semver.Version) {
+	currentVersion := GetVersion()
+	currentVersionSemVer := ConvertVerToSemVar(currentVersion)
+	newerVersionAvailable := currentVersionSemVer.Compare(newestVersionSemVar) == -1
+	var debugFields = log.Fields{"newerVersionAvailable": newerVersionAvailable, "oldVersion": currentVersionSemVer.String(), "newVersion": newestVersionSemVar.String()}
+
+	if newerVersionAvailable {
+		log.WithFields(debugFields).Warnf("There is a newer version of FSOC available, please upgrade from version %s to version %s", currentVersionSemVer.String(), newestVersionSemVar.String())
```
Comment (pnickolov) [inline cmd/version/version.go:116]: Since we already include the versions in the fields, I don't think we need to include them in the message as well. Just say "There is a newer version of fsoc available, please upgrade soon"

File: cmd/version/version.go
Comment (pnickolov) [inline cmd/version/version.go:1]: consider moving this to a versioncheck.go. Even though the file is short, it is quite dense and I think it will be better to separate the code related to the latest-version check.

File: cmd/root.go [hunk 15]
```diff
@@ -222,6 +232,35 @@ func preExecHook(cmd *cobra.Command, args []string) {
...
 	}
+
+	// Do version checking
+	noUpdateWarning, _ := cmd.Flags().GetBool("no-update-warning")
+	updateChecked := int(time.Now().Unix())-getRecentTimestamp() > int(secondInDay) && !noUpdateWarning
+	updateChannel = make(chan *semver.Version)
+	if updateChecked {
+		go version.CheckForUpdate(updateChannel)
+	} else {
+		go func() { updateChannel <- version.ConvertVerToSemVar(version.GetVersion()) }()
```
Comment (pnickolov) [inline cmd/root.go:243]: If the purpose of this is just to not do (skip) the check, isn't there a better way?

**Review by pnickolov — CHANGES_REQUESTED** [review]

**Comment by pavel-georgiev** [comment]
@pnickolov it is great to have the ability to check for available updates. Can we make this to not be the default behavior. This kind of hidden work raises some brows, in most cases would be undesirable and AFAIk it is somewhat unorthodox for an OSS CLI tool. On top of being a bad practice (IMO), I have a few practical cases where it would be undesirable:
1. I do a local operation (like fsoc solution bump), but now I have to wait for a network call for something that I did not ask for
2. Version check fails (or release location changed) - is the exit status going to reflect the status of my operation or the version update?


Here are some suggestions on how to make it less intrusive, ordered by my preference:
* Make this run only as part of `fsoc version` so it can show you your version and latest available version, keep the flag for disabling the remote check
* Make this a "best effort" operation - if it is done by the time the command is done, then tell me about the new version (and/or store the result so you can keep warning me even without making the network call), if it is not done, try next time. In short - never make fsoc run longer because of the version check
* Have a config option to disable this (I suspect it will be a popular option, especially for CI operations and internal)





and rather add a new command for it or only do the check as part of `fsoc version`.

**Review by pnickolov — APPROVED** [review]

**Comment by pnickolov** [comment]
In reply to @pavel-georgiev : I agree with some of the issues but we still expect significant changes and making sure the latest version is in use is important.

I think @GDW1 has addressed most of the issues:
* the check is best effort; if it fails, it will be warned but the operation and its exit code are never affected
* the version check is done in parallel with the main operation
* while very fast operations will still wait for the version check to complete at exit, the check is performed no more often than once a day, so it will not affect most operations (only one per day, less than login)
* there is a global command-line option AND an environment variable to disable the version check (for CI or in environments that don't have access to github or writeable temp dir)
